GPU/Vulkan/ShaderManagerVulkan.cpp

```diff
@@ -0,0 +1,581 @@
+		ERROR_LOG(G3D, "Shader source:\n%s", code);
+#ifdef SHADERLOG
+		OutputDebugStringA("Messages:\n");
+		OutputDebugStringA(errorMessage.c_str());
+		OutputDebugStringA(code);
+#endif
+		Reporting::ReportMessage("Vulkan error in shader compilation: info: %s / code: %s", errorMessage.c_str(), code);
+	} else {
+		success = vulkan_->CreateShaderModule(spirv, &module_);
+#ifdef SHADERLOG
+		OutputDebugStringA("OK");
+#endif
+	}
+
+	if (!success) {
+		failed_ = true;
+		return;
+	} else {
+		DEBUG_LOG(G3D, "Compiled shader:\n%s\n", (const char *)code);
+	}
+}
+
+VulkanFragmentShader::~VulkanFragmentShader() {
+	if (module_) {
+		vulkan_->Delete().QueueDeleteShaderModule(module_);
+	}
+}
+
+std::string VulkanFragmentShader::GetShaderString(DebugShaderStringType type) const {
+	switch (type) {
+	case SHADER_STRING_SOURCE_CODE:
+		return source_;
+	case SHADER_STRING_SHORT_DESC:
+		return FragmentShaderDesc(id_);
+	default:
+		return "N/A";
+	}
+}
+
+VulkanVertexShader::VulkanVertexShader(VulkanContext *vulkan, ShaderID id, const char *code, int vertType, bool useHWTransform) 
+	: vulkan_(vulkan), id_(id), failed_(false), useHWTransform_(useHWTransform), module_(VK_NULL_HANDLE) {
+	source_ = code;
+	std::string errorMessage;
+	std::vector<uint32_t> spirv;
+#ifdef SHADERLOG
+	OutputDebugStringA(code);
+#endif
+	bool success = GLSLtoSPV(VK_SHADER_STAGE_VERTEX_BIT, code, spirv, &errorMessage);
+	if (!errorMessage.empty()) {
+		if (success) {
+			ERROR_LOG(G3D, "Warnings in shader compilation!");
+		} else {
+			ERROR_LOG(G3D, "Error in shader compilation!");
+		}
+		ERROR_LOG(G3D, "Messages: %s", errorMessage.c_str());
+		ERROR_LOG(G3D, "Shader source:\n%s", code);
+		OutputDebugStringUTF8("Messages:\n");
+		OutputDebugStringUTF8(errorMessage.c_str());
+		Reporting::ReportMessage("Vulkan error in shader compilation: info: %s / code: %s", errorMessage.c_str(), code);
+	} else {
+		success = vulkan_->CreateShaderModule(spirv, &module_);
+#ifdef SHADERLOG
+		OutputDebugStringA("OK");
+#endif
+	}
+
+	if (!success) {
+		failed_ = true;
+		module_ = VK_NULL_HANDLE;
+		return;
+	} else {
+		DEBUG_LOG(G3D, "Compiled shader:\n%s\n", (const char *)code);
+	}
+}
+
+VulkanVertexShader::~VulkanVertexShader() {
+	if (module_) {
+		vulkan_->Delete().QueueDeleteShaderModule(module_);
+	}
+}
+
+std::string VulkanVertexShader::GetShaderString(DebugShaderStringType type) const {
+	switch (type) {
+	case SHADER_STRING_SOURCE_CODE:
+		return source_;
+	case SHADER_STRING_SHORT_DESC:
+		return VertexShaderDesc(id_);
+	default:
+		return "N/A";
+	}
+}
+
+static void ConvertProjMatrixToVulkan(Matrix4x4 &in, bool invertedX, bool invertedY) {
+	// Half pixel offset hack
+	float xoff = 0.5f / gstate_c.curRTRenderWidth;
+	xoff = gstate_c.vpXOffset + (invertedX ? xoff : -xoff);
+	float yoff = 0.5f / gstate_c.curRTRenderHeight;
+	yoff = gstate_c.vpYOffset + (invertedY ? yoff : -yoff);
+
+	if (invertedX)
+		xoff = -xoff;
+	if (invertedY)
+		yoff = -yoff;
```

Comment:
Hmm.  Now that I think harder about it, should the half-pixel offset be negated either?  It is also a drift in the final mapping of the pixel to screen coordinates.  It ought to have the same properties...

Also, it's being negated twice in vulkan/d3d9...

-[Unknown]